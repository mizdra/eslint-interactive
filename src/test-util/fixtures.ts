import { exec } from 'node:child_process';
import { randomUUID } from 'node:crypto';
import { rm } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';
import { defineIFFCreator } from '@mizdra/inline-fixture-files';
import fse from 'fs-extra/esm';

const execPromise = promisify(exec);

const rootDir = join(dirname(fileURLToPath(import.meta.url)), '..', '..');

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
export const fixtureDir = join(rootDir, 'tmp/fixture', process.env['VITEST_POOL_ID']!);
export const createIFF = defineIFFCreator({ generateRootDir: () => join(fixtureDir, randomUUID()) });

/**
 * Returns a string containing the stitched together contents of the file modified by fix.
 * To make the snapshot easier to read, the name of the file is inserted at the beginning of the contents of each file.
 */
export async function getSnapshotOfChangedFiles(): Promise<string> {
  const { stdout } = await execPromise(`diff -qr fixtures fixtures-tmp | cut -d " " -f 4 | xargs tail -n +1`, {
    cwd: rootDir,
  });
  return stdout.toString();
}

export async function setupFixturesCopy() {
  await rm(resolve(rootDir, 'fixtures-tmp'), { recursive: true, force: true });
  await fse.copy(resolve(rootDir, 'fixtures'), resolve(rootDir, 'fixtures-tmp'));
}

export async function cleanupFixturesCopy() {
  await rm(resolve(rootDir, 'fixtures-tmp'), { recursive: true, force: true });
}
