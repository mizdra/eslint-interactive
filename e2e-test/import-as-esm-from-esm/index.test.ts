import { afterEach, beforeEach, expect, test } from 'vitest';
import { execSync } from 'child_process';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import stripAnsi from 'strip-ansi';
// eslint-disable-next-line no-restricted-imports
import { cleanupFixturesCopy, getSnapshotOfChangedFiles, setupFixturesCopy } from '../../src/test-util/fixtures.js';

const { Core, takeRuleStatistics } = await import('eslint-interactive');

beforeEach(async () => {
  await setupFixturesCopy();
});

afterEach(async () => {
  await cleanupFixturesCopy();
});

test('Programmable API', async () => {
  const core = new Core({
    patterns: ['fixtures-tmp'],
    eslintOptions: {
      type: 'legacy',
      cwd: join(dirname(fileURLToPath(import.meta.url)), '..', '..'),
    },
  });
  const results = await core.lint();

  expect(stripAnsi(core.formatResultSummary(results))).toMatchSnapshot();
  const statistics = takeRuleStatistics(results);
  expect(statistics).toMatchSnapshot();

  await core.applyAutoFixes(results, ['semi']);
  expect(await getSnapshotOfChangedFiles()).toMatchSnapshot();
});
