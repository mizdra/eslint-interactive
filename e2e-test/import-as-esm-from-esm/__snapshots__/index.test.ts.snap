// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Programmable API 1`] = `
"- 8 files (0 file passed, 8 files failed) checked.
- 57 problems (56 errors, 1 warning) found.
╔═════════════════════════════╤═══════╤═════════╤════════════╤═════════════════╗
║ Rule                        │ Error │ Warning │ is fixable │ has suggestions ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ ban-exponentiation-operator │ 8     │ 0       │ 0          │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ semi                        │ 12    │ 1       │ 13         │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ import/order                │ 3     │ 0       │ 3          │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ prefer-const                │ 4     │ 0       │ 4          │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ no-unused-vars              │ 7     │ 0       │ 0          │ 0               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ no-useless-escape           │ 5     │ 0       │ 0          │ 5               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ no-unsafe-negation          │ 5     │ 0       │ 0          │ 5               ║
╟─────────────────────────────┼───────┼─────────┼────────────┼─────────────────╢
║ arrow-body-style            │ 12    │ 0       │ 12         │ 0               ║
╚═════════════════════════════╧═══════╧═════════╧════════════╧═════════════════╝
"
`;

exports[`Programmable API 2`] = `
[
  {
    "errorCount": 8,
    "hasSuggestionsCount": 0,
    "hasSuggestionsErrorCount": 0,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 0,
    "isFixableErrorCount": 0,
    "isFixableWarningCount": 0,
    "ruleId": "ban-exponentiation-operator",
    "warningCount": 0,
  },
  {
    "errorCount": 12,
    "hasSuggestionsCount": 0,
    "hasSuggestionsErrorCount": 0,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 13,
    "isFixableErrorCount": 12,
    "isFixableWarningCount": 1,
    "ruleId": "semi",
    "warningCount": 1,
  },
  {
    "errorCount": 3,
    "hasSuggestionsCount": 0,
    "hasSuggestionsErrorCount": 0,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 3,
    "isFixableErrorCount": 3,
    "isFixableWarningCount": 0,
    "ruleId": "import/order",
    "warningCount": 0,
  },
  {
    "errorCount": 4,
    "hasSuggestionsCount": 0,
    "hasSuggestionsErrorCount": 0,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 4,
    "isFixableErrorCount": 4,
    "isFixableWarningCount": 0,
    "ruleId": "prefer-const",
    "warningCount": 0,
  },
  {
    "errorCount": 7,
    "hasSuggestionsCount": 0,
    "hasSuggestionsErrorCount": 0,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 0,
    "isFixableErrorCount": 0,
    "isFixableWarningCount": 0,
    "ruleId": "no-unused-vars",
    "warningCount": 0,
  },
  {
    "errorCount": 5,
    "hasSuggestionsCount": 5,
    "hasSuggestionsErrorCount": 5,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 0,
    "isFixableErrorCount": 0,
    "isFixableWarningCount": 0,
    "ruleId": "no-useless-escape",
    "warningCount": 0,
  },
  {
    "errorCount": 5,
    "hasSuggestionsCount": 5,
    "hasSuggestionsErrorCount": 5,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 0,
    "isFixableErrorCount": 0,
    "isFixableWarningCount": 0,
    "ruleId": "no-unsafe-negation",
    "warningCount": 0,
  },
  {
    "errorCount": 12,
    "hasSuggestionsCount": 0,
    "hasSuggestionsErrorCount": 0,
    "hasSuggestionsWarningCount": 0,
    "isFixableCount": 12,
    "isFixableErrorCount": 12,
    "isFixableWarningCount": 0,
    "ruleId": "arrow-body-style",
    "warningCount": 0,
  },
]
`;

exports[`Programmable API 3`] = `
"==> fixtures-tmp/index.js <==
import { mul } from './lib/mul';
import { add } from './lib/add';
import { sub } from './lib/sub';
import { exp } from './lib/exp';
import { strictEqual } from 'assert';

const numA=add(1, 2);
let numB=add(3, 4);
let numC=sub(5, 6);
let numD=exp(2, 2);
let numE=mul(numA, numB, numC, numD);

strictEqual(numD, -84);

==> fixtures-tmp/lib/shebang.js <==
#!/usr/bin/env node

console.log();

==> fixtures-tmp/lib/warning.js <==
/* eslint semi: 1 */

export function add(a, b) {
  return a + b;
}
"
`;
